[feat] Waveform & Score Title 기능 구현

ehdrbdndns opened the pull request:
### 설명
ChordProgress View의 Waveform 컴포넌트와 Score Title을 수정하는 기능을 개발했습니다.

### 관련 이슈

Closes #40

### 작업 내용
- [x] Waveform UI 개발
- [x] 음악 재생 시, Waveform 애니메이션 동작
- [x] Waveform 탭할 시 해당 시간대로 오디오 이동
- [x] Score Title 수정 기능 개발

### 스크린샷

https://github.com/user-attachments/assets/c6d4a871-45f3-4ee9-968e-a209d1631376

 (Optional)


### 참고 내용

ehdrbdndns committed:
[refactor] Ensures model is available before use

ehdrbdndns committed:
[feat] Adds waveform view

users to visually understand audio amplitude over time.

This change introduces a new `Waveform` view that displays
audio levels as a series of capsules, with height
corresponding to amplitude.

feat/#40

ehdrbdndns committed:
[feat] Adds audio levels to the score

ehdrbdndns committed:
[feat] Adds waveform view to ChordProgressVIew

ehdrbdndns committed:
[refactor] Score mock data

ehdrbdndns committed:
[#40] feat: waveform light/dark mode

ehdrbdndns committed:
[#40] feat: seek audio file from timesignatureEnables seeking to specific time

ehdrbdndns committed:
[feat]: waveform 선택 시 오디오 이동

ehdrbdndns committed:
[refactor] geometryReader 간소화

ehdrbdndns committed:
[#40] feat: prevent code text overflow

ehdrbdndns committed:
[#40] feat: update score title

ehdrbdndns committed:
[#40] feat: edit navigation title

ehdrbdndns committed:
[#40] feat: add gitignore

ehdrbdndns committed:
[#40] feat: TimeRuler mask 처리

Segyun commented on GMG/Core/Utils/View+Extensions.swift:
```diff
@@ -13,12 +13,19 @@ extension View {
+            ) {
                 leading()
             } trailing: {
                 trailing()
```
`NavigationBar` 자체에 타이틀 변경 로직을 넣기보다는 `Center`로 뷰를 넘겨주는 식으로 하는 것이 어떨까요? 확장성 면에서 더 좋을 것 같습니다!

```suggestion
extension View {
    func navigationBar<Leading: View, Center: View, Trailing: View>(
        _ title: String? = nil,
        isBackButtonHidden: Bool = false,
        @ViewBuilder leading: () -> Leading,
        @ViewBuilder center: () -> Center,
        @ViewBuilder trailing: () -> Trailing
    ) -> some View {
    VStack {
        NavigationBar(
            title,
            isBackButtonHidden: isBackButtonHidden,
            isTitleEditable: isTitleEditable,
            onEnterTitle: onEnterTitle
        ) {
            leading()
        } center: {
            center()
        } trailing: {
            trailing()
```

ehdrbdndns replied:
네 확장성 말고도, 통일성 측면에서도 leading이랑 trailing 처럼 컴포넌트를 인자로 받는게 좋겠네요!

ehdrbdndns replied:
네 확장성 말고도, 통일성 측면에서도 leading이랑 trailing 처럼 컴포넌트를 인자로 받는게 좋겠네요!

Segyun commented on GMG/Scenes/ChordProgress/ChordProgressView.swift:
Text 뷰의 값을 없애는 것보다 Text 뷰 자체를 없애거나 투명하게 만드는 것은 어떨까요? 그리고 `showText` 보다는 `showChordDescription`과 같이 의미를 상세히 적어주는 것도 좋을 것 같습니다!

```suggestion
                    ZStack {
                        if showText {
                            Text(chord.description)
                                .font(Typography.WantedSansStd.R7)
                                .foregroundStyle(
                                    foregroundColor
                                )
                        }
                    }
```

```suggestion
                    Text(chord.description)
                        .font(Typography.WantedSansStd.R7)
                        .foregroundStyle(
                            foregroundColor
                        )
                        .opacity(showText ? 1.0 : 0.0)
```

ehdrbdndns replied:
코드 스타일 통합 좋습니다~!

Segyun commented on GMG/Scenes/ChordProgress/Waveform.swift:
```diff
@@ -0,0 +1,181 @@
+                            width: capsuleWidth,
+                            height: capsuleHeight(for: amplitude)
+                        )
+                }
```
```suggestion
                ForEach(Array(amplitudes.enumerated()), id: \.0) { index, amplitude in
                    Capsule()
                        .fill(color)
                        .frame(
                            width: capsuleWidth,
                            height: capsuleHeight(for: amplitude)
                        )
                }
```

ehdrbdndns replied:
버그 방지 코드 좋습니당

Segyun commented on GMG/Scenes/ChordProgress/Waveform.swift:
```diff
@@ -0,0 +1,181 @@
+        } else {
+            return resample(amplitudes, to: targetCount)
+        }
+    }
```
```Swift
    private var capsuleCount: Int { capsuleCount(for: width) }
    private var preparedAmplitudes: [Float] { amplitudesForRendering(targetCount: capsuleCount) }
    private var progressWidth: CGFloat { fillWidth(totalWidth: width) }
```

프로퍼티 값을 메서드로 분리한 이유는 길어서인가요? 재사용되지 않는다면 프로퍼티 안에 넣는 것도 좋을 것 같아요!

ehdrbdndns replied:
그렇네요! 재사용되지 않으니까 프로퍼티로 빼는게 좋겠네요~!

Segyun commented on GMG/Scenes/ChordProgress/Waveform.swift:
```diff
@@ -0,0 +1,181 @@
+            capsuleSpacing: CGFloat = 4,
+            horizontalPadding: CGFloat = 11,
+            minCapsuleHeight: CGFloat = 6,
+            maxCapsuleHeight: CGFloat = 21
```
기본 값 지정은 빼도 될 것 같아요!

Segyun commented on GMG/Scenes/ChordProgress/Waveform.swift:
```diff
@@ -0,0 +1,181 @@
+        }
+    }
+    
+    private var capsuleCount: Int { capsuleCount(for: width) }
+    private var preparedAmplitudes: [Float] { amplitudesForRendering(targetCount: capsuleCount) }
+    private var progressWidth: CGFloat { fillWidth(totalWidth: width) }
+    
+    var body: some View {
+        HStack {
+            ZStack(alignment: .leading) {
+                RoundedRectangle(cornerRadius: 12)
+                    .fill(backgroundColor)
+                
+                CapsuleStack(
+                    amplitudes: preparedAmplitudes,
+                    color: unFilledCapsuleColor
+                )
+                
+                CapsuleStack(
+                    amplitudes: preparedAmplitudes,
+                    color: filledCapsuleColor
+                )
```
매개변수 전달이 빠진 것 같아요!

```Swift
            capsuleWidth: capsuleWidth,
            capsuleSpacing: capsuleSpacing,
            horizontalPadding: horizontalPadding,
            minCapsuleHeight: minCapsuleHeight,
            maxCapsuleHeight: maxCapsuleHeight
```

Segyun commented on GMG/Scenes/ChordProgress/Waveform.swift:
```diff
@@ -0,0 +1,181 @@
+            }
+            
+            let t = scaled - Float(lower)
+            return source[lower] * (1 - t) + source[upper] * t
```
`t`의 의미를 명확하게 알 수 있도록 이름 바꾸는 것은 어떨까요?

```suggestion
            let interpolationFactor = scaled - Float(lower)
            return source[lower] * (1 - interpolationFactor) + source[upper] * interpolationFactor
```

ehdrbdndns replied:
좋습니다~!

Segyun commented on GMG/Scenes/ChordProgress/Waveform.swift:
```diff
@@ -0,0 +1,181 @@
+                    amplitudes: preparedAmplitudes,
+                    color: filledCapsuleColor
+                )
+                .mask(
+                    HStack(spacing: 0) {
+                        RoundedRectangle(cornerRadius: 12)
+                            .frame(width: progressWidth)
+                        Spacer(minLength: 0)
+                    }
+                )
```
마스크로 가리는 것보다 n번 째 캡슐까지 색상을 채우는 방식으로 하는 것은 어떨까요? 그러면 하나의 CapsuleStack만 사용해도 될 것 같아요.

ehdrbdndns replied:
애니메이션 효과를 주려고 .mask를 사용했습니다! 단순히 캡슐 색상을 채우는 것보다 왼쪽에서 오른쪽으로 색상이 스르륵 채워지는 효과가 이쁘더리구요.

Segyun commented on GMG/Scenes/Recording/RecordingModel.swift:
```diff
@@ -100,6 +100,7 @@ final class RecordingModel:
     }
     
     func finishScoreCreation(_ score: Score) {
+        score.audioLevels = audioLevels
```
이 부분은 추후에 `ScoreFactory`로 옮기도록 하겠습니다. 지금 `audioLevels`는 재생 시에 초기화되어서 모든 값을 가지고 있지 않을 수도 있어서 이 값을 그대로 사용하면 안됩니다!

ehdrbdndns replied:
WaveForm에 들어갈 진폭 리스트는 녹음할 때 추출한 값을 사용하는게 좋을 것 같아요! 혹시 진폭 배열을 담는 다른 값들이 있나요?

Segyun replied:
처음에는 Score에 audioLevels를 저장하는 것 대신에 `ChordProgressView`에서 오디오 파일로 변환하는 것을 생각하긴 했었습니다! (앱에서 녹음하지 않고 오디오 파일 불러오는 기능이 추가된다면)

그럼 일단 `RecordingView`에서 전체 Audio Level을 저장하고 재생할 때는 이를 활용하는 방식으로 로직을 수정해보겠습니다.

ehdrbdndns replied:
네, 좋습니다!

다만, 사용자가 Waveform에 표시되는 정보들을 아주 정확하게 확인해야 할 필요가 없다보니, 앱 기능 자체에는 모든 값이 완전하게 존재하지 않더라도 큰 문제는 없을 것 같습니다~!

ehdrbdndns replied:
후순위로 미루는게 좋겠어요!

Segyun commented on .gitignore:
```diff
@@ -57,3 +57,5 @@ fastlane/report.xml
 fastlane/screenshots/**/*.png
 fastlane/test_output
+
+GMG.xcodeproj/xcshareddata/xcschemes/GMG.xcscheme
```
이 스킴은 공유되어야 하는데 프로젝트 옮기면서 빠진 듯합니다. `.gitignore`에 넣지 말고 새로 이슈 만들어서 다시 올리겠습니다!

ehdrbdndns replied:
넵!

Segyun commented on GMG/UI/Components/NavigationBar.swift:
```diff
@@ -3,50 +3,126 @@
 import SwiftUI
 
 struct NavigationBar<Leading: View, Trailing: View>: View {
```
`View+Extensions.swift`에서 말한 것처럼 타이틀 변경을 `Center: View`로 받아서 만드는 것이 좋을 것 같습니다~

중간에 `title`에 따라 분기처리하면 될 것 같아요!

```Swift
if let title {
    Text(title)
} else {
    center
}
```

그리고 지금 HStack으로 Spacer만 처리하면은 Trailing 뷰가 없으면 가운데 뷰가 오른쪽으로 치우쳐져서 ZStack 구조 유지 부탁드립니다!

ehdrbdndns replied:
center를 받게 되면, title을 따로 받을 필요가 없어서 매개변수에서 제외시켰습니다~!

Segyun reviewed (changes requested):
수고하셨습니다~

ehdrbdndns committed:
[#40] refactor: navigation bar re struct

ehdrbdndns committed:
[#40] refactor:  renames showText to showChordDescription

ehdrbdndns committed:
[#40] refactor: filters chord cells by duration ratio

ehdrbdndns committed:
[#40] refacor: Improves waveform rendering and resampling

ehdrbdndns committed:
[#40] chore: gitignore

ehdrbdndns committed:
[#40] chore: dev to feat/#40

Segyun committed:
[#40] chore: resolve `ScorePlayer` conflict

Segyun commented on GMG/Scenes/ChordProgress/ChordProgressView.swift:
```diff
@@ -603,39 +759,40 @@ extension ChordProgressView {
-                    Text(chord.description)
-                        .font(Typography.WantedSansStd.R7)
-                        .foregroundStyle(
-                            foregroundColor
-                        )
-                        .frame(
-                            maxWidth: .infinity,
-                            maxHeight: .infinity
-                        )
-                        .background(
-                            backgroundColor,
-                            in: RoundedRectangle(cornerRadius: 12)
-                        )
+                    ZStack {
+                        if showChordDescription {
+                            Text(chord.description)
+                                .font(Typography.WantedSansStd.R7)
+                                .foregroundStyle(
+                                    foregroundColor
+                                )
+                                .frame(
+                                    maxWidth: .infinity,
+                                    maxHeight: .infinity
+                                )
+                                .background(
+                                    backgroundColor,
+                                    in: RoundedRectangle(cornerRadius: 12)
+                                )
+                        } else {
+                            VStack {}
+                                .frame(
+                                    maxWidth: .infinity,
+                                    maxHeight: .infinity
+                                )
+                                .background(
+                                    backgroundColor,
+                                    in: RoundedRectangle(cornerRadius: 12)
+                                )
+                        }
+                    }
```
이렇게 구현하자는 의미였어요!

```suggestion
                    ZStack {
                        if showChordDescription {
                            Text(chord.description)
                                .font(Typography.WantedSansStd.R7)
                                .foregroundStyle(
                                    foregroundColor
                                )
                        }
                    }
                    .frame(
                        maxWidth: .infinity,
                        maxHeight: .infinity
                    )
                    .background(
                        backgroundColor,
                        in: RoundedRectangle(cornerRadius: 12)
                    )
```

ehdrbdndns replied:
좋습니다!

Segyun commented on GMG/UI/Components/NavigationBar.swift:
```diff
@@ -2,58 +2,53 @@
 
-                trailing
-            }
-            .buttonStyle(NavigationBarButtonStyle())
-
-            if let title {
-                Text(title)
-                    .font(Typography.WantedSansStd.R6)
-                    .foregroundStyle(
-                        colorScheme == .light
-                            ? Color.black1 : Color.white1
-                    )
-            }
+            center
+
+            Spacer()
+
+            trailing
         }
         .padding(Spacing.md)
+        .buttonStyle(NavigationBarButtonStyle())
     }
```
HStack을 쓰면 trailing이 없으면 center가 오른쪽으로 치우치게 됩니다. ZStack으로 수정해주세요.

<p align="center">
  <img width="256" alt="NavigationBar" src="https://github.com/user-attachments/assets/590a2e3f-b642-43b5-8047-ab880c2cf610" />
</p>

```suggestion
    var body: some View {
        ZStack {
            HStack {
                if !isBackButtonHidden {
                    BackButton()
                }
                
                leading
                
                Spacer()
                
                trailing
            }
            
            center
        }
        .padding(Spacing.md)
        .buttonStyle(NavigationBarButtonStyle())
    }
```

ehdrbdndns replied:
오 그렇네요! 중간이랑 양 옆의 레이아웃을 분리해서 넣는 방식이 다소 낯설긴한데 center를 정확하게 중간에 두는 좋은 방법이네요!

Segyun commented on GMG/Scenes/ChordProgress/ChordProgressView.swift:
```diff
@@ -108,6 +114,93 @@ struct ChordProgressView: View {
+            DispatchQueue.main.async {
+                isTitleFieldFocused = true
+            }
+        }
```
`startTitleEditing`과 `finishEditingTitle`로 명확히 구분하고 호출부에서 분기 처리하는 것이 좋을 것 같습니다.

```suggestion
        private func startTitleEditing() {
            isTitleEditing = true
            DispatchQueue.main.async {
                isTitleFieldFocused = true
            }
        }
```

ehdrbdndns replied:
네 finishEditingTitle이 있으니, toggle보다는 start로 분리하는게 좋겠네요. 클린코드적으로도 맞는 방향인 것 같습니다! 감사합니다~!

Segyun commented on GMG/Scenes/ChordProgress/ChordProgressView.swift:
```diff
@@ -459,33 +594,32 @@ extension ChordProgressView {
                                         chord: chord,
-                                        isCurrentChord: currentChordCell?
-                                            .startTime
+                                        showChordDescription: widthRatio >= 0.2,
```
0.02인데 오타난 것 같아요!

<p align="center">
  <img width="256" alt="ChordProgressView Segment" src="https://github.com/user-attachments/assets/69b8e016-39b5-40b3-a4a7-74c48d4e7d71" />
</p>

```suggestion
                                        showChordDescription: widthRatio >= 0.02,
```

ehdrbdndns replied:
이거는 텍스트가 보이는지 말지에 대한 비율이라 20%가 맞습니다! 아리랑 이야기된 부분이에요!

ehdrbdndns replied:
다소 헷갈리는 부분이 있을 수 있으니 코드를 좀 수정해볼게용

nakisara01 reviewed (dismissed):
열심히 봤는데 제 레벨에서 고칠 것 없는 것 같네용.. 수고하셨슴당
레몬이 달아주신 것들 중에 반영해서 머지하면 될 것 같아요!

ehdrbdndns committed:
[#40] refactor: re struct of navigation layout

ehdrbdndns committed:
[#40] refactor: rename variable

display and interaction within a specified time segment. It
introduces a new data structure for managing chord entries
within segments, calculates occupancy ratios for display,
and optimizes the visibility of chord descriptions.

feat/#40

ehdrbdndns committed:
[#40] chore

ehdrbdndns committed:
[#40] feat: add text auto fit

ehdrbdndns commented:
아래 느낌으로 글꼴이 줄어듭니다~!


<img width="442" height="931" alt="image" src="https://github.com/user-attachments/assets/b1221563-4f1a-412a-9539-08b9d6675db4" />

Segyun committed:
Merge branch 'dev' into feat/#40

Segyun committed:
[#40] refactor: remove unused code in `navigationBar`
